### PR TITLE
Use nixpkgs stdenv

### PR DESCRIPTION
**Pull request by balsoft**
Use nixpkgs stdenv for setting up environment variables.

Fixes https://github.com/nickel-lang/nickel-nix/issues/72

### CODE REVIEW

File: builders.ncl
Comment (yannham) [inline builders.ncl:45]: It's customary to use the simplest delimiter for multiline strings, `m%"`. Longer delimiters are mostly used when you have other Nickel mutliline strings or escaped interpolation sequences inside (e.g. when documenting a stdlib function, and you want to include a Nickel example which itself has a mutliline string) and don't want to escape stuff too much. But here, it's just text

```suggestion
    | doc m%"
```

File: builders.ncl
Comment (yannham) [inline builders.ncl:83]:
```suggestion
    | doc m%"
```

File: builders.ncl
Comment (yannham) [inline builders.ncl:48]:
```suggestion
    "%
```

File: builders.ncl
Comment (yannham) [inline builders.ncl:86]:
```suggestion
    "%
```

File: builders.ncl
Comment (yannham) [inline builders.ncl:3]: In general we tend to use snake_case for identifiers, and CamelCase for types/contracts.

```suggestion
let concat_strings_sep = fun sep values =>
```

File: builders.ncl
Comment (yannham) [inline builders.ncl:100]: Same thing about casing. Unless those are values that are then exported to Nix? Even there, I wonder if we shouldn't do the translation by hand between snake_case and camelCase. Otherwise it's going to be really confused as to which casing to use. Related discussion: https://github.com/tweag/nickel/discussions/1311

Comment (balsoft) [reply builders.ncl:100]: This is passed to nixpkgs stdenv: https://nixos.org/manual/nixpkgs/stable/#var-stdenv-buildInputs .

Comment (yannham) [reply builders.ncl]: I see, we'll have to think about this at some point, but this PR might not be the right scope. If it's passed onto Nixpkgs, we don't have much choice, at least for now.

File: builders.ncl
Comment (yannham) [inline builders.ncl:8]: I missed that on the first review, but you can use `std.array.reduce_left` for folding starting with the first element, without having to `drop_first` manually

**Review by yannham — APPROVED** [review]
LGTM, as far as I can tell on the Nickel side :+1:  it's nice to see that provided shells also seem less boilerplaty.

File: builders.ncl
Comment (yannham) [reply builders.ncl:8]: https://nickel-lang.org/stdlib/std-array/#reduce_left

Comment (balsoft) [reply builders.ncl]: Thanks, done!

**Comment by yannham** [comment]
@balsoft Should I just merge this, or do I let you push the green button? Or do you prefer a second review from @thufschmitt ?

**Comment by balsoft** [comment]
I would love a second review, as it's quite a big change.

File: builders.ncl
Comment (thufschmitt) [reply builders.ncl:74]: Actually, using `structuredAttrs` mean that we can directly pass a record as `buildInputs` (it'll be passed as an associative array rather than a numerically-indexed array, but that doesn't change anything). So we can get rid of the concat dance altogether

File: builders.ncl
Comment (thufschmitt) [inline builders.ncl:99]: There's a bit of confusion here between what's an input and what's an output. But it's already the case atm, so it's more a note for later than a real concern with this PR

**Review by thufschmitt — CHANGES_REQUESTED** [review]
I like that change quite a lot. I'd like to try using `__structuredAttrs` before merging as it feels like it could make the implementation cleaner though

File: builders.ncl
Comment (balsoft) [reply builders.ncl:74]: > Could we instead use structuredAttrs? stdenv is supposed to be able to handle that just fine now, and that would be nicer than doing this kind of concatenation ourselves

It gets annoying fast. Since `nix-s%"..."%` gets translated to a record (`{ prefix = ..., tag = ..., fragments = ... }`) internally, we have to distinguish between that sort of record and actual records of nix strings. And when I did that, I got `error: boost::bad_format_string: format-string is ill-formed` from Nix :D

Comment (thufschmitt) [reply builders.ncl:74]: OK, let's keep away from it for now then.

> I got error: boost::bad_format_string: format-string is ill-formed from Nix :D

:partying_face:

Comment (balsoft) [reply builders.ncl]: Actually, figured it out. I've kept `env` and `structured_env`.

**Comment by thufschmitt** [comment]
#96 fixed the “use nixpkgs stdenv” part, so the remaining bit here is the `WithInput` shells (which automatically add their inputs to `buildInputs`), and these will be superseded by #100. So closing this. @balsoft feel free to reopen if there's something more to pick from this PR that I've missed

**Pull request closed without merging**
